models/marts/dim_contract_data_hist.sql

```diff
@@ -0,0 +1,216 @@
+    , date_chained as (
+        select
+            cdc.*
+            , lead(cdc.start_date) over (partition by cdc.ledger_key_hash order by cdc.start_date) as next_start_date
+            -- Operation Types:
+            -- INSERT_NEW_KEY: First occurrence of a ledger_key_hash
+            -- START_NEW_VERSION: New version of existing record due to attribute changes
+            -- END_CURRENT_VERSION: Close current version due to future changes
+            -- KEEP_CURRENT: No changes needed, maintain current record
+            , case
+                when cdc.change_type = 'Insert' then 'INSERT_NEW_KEY'
+                when cdc.change_type = 'Update' then 'START_NEW_VERSION'
+                when
+                    cdc.change_type = 'NoChange'
+                    and lead(cdc.start_date) over (partition by cdc.ledger_key_hash order by cdc.start_date) is not null
+                    then 'END_CURRENT_VERSION'
+                else 'KEEP_CURRENT'
+            end as operation_type
+        from combined_cdc_data as cdc
+    )
+
+    -- Final data processing with all transformations
+    , final_data as (
+        select
+            dc.ledger_key_hash
+            , coalesce(dc.source_contract_id, dc.target_contract_id) as contract_id
+            , coalesce(dc.source_contract_durability, dc.target_contract_durability) as contract_durability
+            , coalesce(dc.source_asset_code, dc.target_asset_code) as asset_code
+            , coalesce(dc.source_asset_issuer, dc.target_asset_issuer) as asset_issuer
+            , coalesce(dc.source_asset_type, dc.target_asset_type) as asset_type
+            , coalesce(dc.source_balance, dc.target_balance) as balance
+            , coalesce(dc.source_balance_holder, dc.target_balance_holder) as balance_holder
+            , coalesce(dc.source_contract_create_ts, dc.target_contract_create_ts) as contract_create_ts
+            , coalesce(dc.source_contract_delete_ts, dc.target_contract_delete_ts) as contract_delete_ts
+            , coalesce(dc.source_closed_at, dc.target_closed_at) as closed_at
+            , dc.start_date
+            , coalesce(date_sub(dc.next_start_date, interval 1 day), date('9999-12-31')) as end_date
+            , coalesce(row_number() over (partition by dc.ledger_key_hash order by dc.start_date desc) = 1, false) as is_current
+            , coalesce(dc.source_airflow_start_ts, dc.target_airflow_start_ts) as airflow_start_ts
+            , coalesce(dc.source_batch_id, dc.target_batch_id) as batch_id
+            , coalesce(dc.source_batch_run_date, dc.target_batch_run_date) as batch_run_date
+            , coalesce(dc.source_row_hash, dc.target_row_hash) as row_hash
+            , coalesce(dc.target_dw_load_ts, current_timestamp()) as dw_load_ts
+            , current_timestamp() as dw_update_ts
+            , dc.operation_type
+        from date_chained as dc
+        where dc.operation_type in ('INSERT_NEW_KEY', 'START_NEW_VERSION', 'END_CURRENT_VERSION')
+    )
```

Comment:
Is there a need to actually do the coalescing between the source and target? I would say that any time there is a new `ledger_key_hash` + `closed_at` that is newer should be inserted directly as is even if the only diff is the `closed_at` column

Reply:
- Since we are not tracking all fields/json changes, a change in closed_at doesn't always necessarily result in a new row. Only a change in the fields considered for CDC (row_hash calculation) can trigger a change. Closed_at is not part of row_hash calculation

- This block necessarily combines fetching the target record (for closing previous version) as well as fetching the source records (New Inserts or Updates to existing keys) into one CTE.

Reply:
I would argue that `closed_at` should trigger a new scd2 row even if other columns don't change. Because there is a different `closed_at` means that a ledger_key_hash had a new ledger entry. If we don't include that change we wouldn't be able to use the scd table to find all the instances of a ledger_key_hash entry in time

Reply:
sure. In the current state, I didn't find any use because of the fields we are using for CDC tracking and tracking every record will look like a duplicate in the SCD2 table if the CDC fields doesn't change. If we properly track the actual change in the json fields as well, then I totally agree with the suggestion. As it stands, it is going to look like a temporal chain of records that get closed and opened without any materialistic changes in the data. 

I will table this for a deeper discussion during office hours as well